Lib/test/datetimetester.py

```diff
@@ -2919,6 +2944,14 @@ def test_strftime(self):
         # A naive object replaces %z and %Z with empty strings.
         self.assertEqual(t.strftime("'%z' '%Z'"), "'' ''")
 
+        # bpo-34482: Check that surrogates don't cause a crash.
+        # FIXME: The C datetime implementation raises an exception
+        # while the pure-Python one succeeds.
+        try:
+            t.strftime('\ud800')
```

Comment:
@pganssle, why assert that? It's not clear to me that this should be something we ensure.

IMO the test is fine. I'd just suggest removing the "FIXME" paragraph entirely (here and above).

Reply:
Are we saying that `.strftime(unicode_with_surrogates)` is *undefined behavior*? Implementation dependent?

I think it's reasonable to either pick something and stick with it or formally make this an error on the pure-python side as well.

Reply:
Sounds good to me. Let's go with your fix (in the PR to this PR's branch) and the test you suggest above. @izbyshev, could you update accordingly?

Reply:
Hm.. This seems like a separate bug on Mac, honestly, rather than implementation-dependent behavior. Usually the "implementation-dependent" stuff in strftime is in how different platforms interpret the different formatting directives, not stuff like this.

I'd be tempted to move the surrogate *assertion* portion added into a separate test that you can wrap in `@unittest.skipIf` or the like.

As for what to do about it, technically it's probably a bug in the platform's `strftime` implementation, of course, but it might be worth adding a workaround like I did for the C implementation into the Python implementation. It definitely seems wrong that `dt.strftime('%Y-%m-%dT%H:%M:%S\ud800')` would return `''`. It should either raise an error or return the formatted date with the surrogate character in place.

Reply:
It turned out that behavior of `strftime` with surrogates is even [worse](https://github.com/python/cpython/blob/ef8861c112ed1dac9351958c121bc24ca4ecdb08/Modules/timemodule.c#L731). If there is no `wcsftime`, the string is encoded with `surrogateescape` error handler, which means that my tests should fail with `UnicodeEncodeError` even on pure-Python `datetime` because the surrogate I use (`\uD800`) is not in the range of surrogates used for escaping (`\uDC80-\uDCFF`).

This is indeed what happens on my Windows 8.1 with Python 3.7 in a manual test:
```
Python 3.7.0 (v3.7.0:1bf9cc5093, Jun 27 2018, 04:59:51) [MSC v.1914 64 bit (AMD64)] on win32
>>> import time
>>> time.strftime('\ud800')
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
UnicodeEncodeError: 'locale' codec can't encode character '\ud800' in position 0: encoding error
```

Moreover:
```
>>> time.strftime('\udc80')
'\x80'
```

This I can't explain currently. Decoding in timemodule.c uses the locale encoding (with `surrogateescape`), and my locale's encoding is not Latin-1. If I do what I think should be the same, I get a different result:
```
>>> ord('\udc80'.encode('mbcs', 'surrogateescape').decode('mbcs', 'surrogateescape'))
1026 # This is the unicode code point corresponding to b'0x80' in cp1251 encoding
```

I'll try to understand this tomorrow.

Reply:
Here is the summary of what I've learned.
* [`strftime` implementation](https://github.com/python/cpython/blob/ef8861c112ed1dac9351958c121bc24ca4ecdb08/Modules/timemodule.c#L679) uses either `strftime` or `wcsftime` from the C standard library, depending on the platform (usually `wcsftime` is available, e.g. on Linux and macOS, but `strftime` is used on Windows because `wcsftime` is broken).
* Those functions are locale-dependent (in the sense of `setlocale()`).
* Those functions don't report errors via `errno` according to [POSIX](http://pubs.opengroup.org/onlinepubs/9699919799/functions/wcsftime.html) and man pages. All they can do is return 0, except that on Windows `errno` is set to `EINVAL` if buffer is too short.
* If `wcsftime` is used, Python converts the format string into `wchar_t` string, and there are no issues with surrogates at this point. However, `wcsftime` may perform additional validation of wchars for the current locale. `glibc` apparently doesn't do that, since I could roundtrip any code point through `wcsftime`, including surrogates. But `wcsftime` on macOS does perform the validation: in C locale, it rejects code points > 255 (by returning zero; `errno` was also set to `EILSEQ` in my tests, but it's not documented). Even in a locale with UTF-8 charset, it rejects surrogates. This why tests on macOS failed.
* If `strftime` is used, Python encodes the input into a locale-dependent charset with `surrogateescape` error handler, and decodes the result back in the same way. The implications:
  * Surrogates outside of `\uDC80-\uDCFF` range cause `UnicodeEncodeError`.
  * Other surrogates can't be reliably round-tripped because they turn into lone bytes on encoding which are then reinterpreted with the locale encoding on the way back. This explains why I got `0x80` from `time.strftime('\udc80')` on Windows: on encoding, the surrogate was unwrapped into `b'0x80'`, on decoding, since there was no `setlocale()` calls and the C locale was active, each byte was simply converted to the code point with the same numerical value. I was initially confused because I thought that "locale" means the user-default code page (which would give a different interpretation for `b'0x80'`).

Given the above, I don't think that we should check the return value of `strftime` in surrogate-related tests, at least in this PR. If such tests are desired, I think that they belong to the `time` module and require some experiments/doc research on platforms not covered by build bots.

Reply:
@izbyshev Awesome work. I'm thinking the best way forward at this point is to drop the assertion portion and to copy all this information into a separate bug report about the inconsistency in strftime. It might be a `wontfix` or back-burnered, but I think we can probably get a lot more consistency than we already have.

Reply:
@pganssle After thinking about a separate bug report, I realized that the whole situation is not so different from the usual stuff with file-system-related functions. Some platforms have wide-character APIs, others have only byte-oriented APIs, and Python tries to appease both. `open('\ud800')` fails on Linux with `UnicodeEncodeError`, but is OK on Windows. And the twist with `wcsftime` on macOS rejecting surrogates is somewhat similar Windows refusing to `open('*')`. Seems unlikely that we can reach consistency here, given that existing users may rely on specific behavior. So I've submitted bpo-34512 to at least try to update the docs.